internal/aws/xray/telemetry/recorder.go

```diff
@@ -0,0 +1,327 @@
+func (t *telemetryRecorder) start() {
+	ticker := time.NewTicker(t.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if t.recordUpdated.Load() {
+				t.add(t.cutoff())
+				t.recordUpdated.Store(false)
+				if failedToSend, err := t.send(t.flush()); err != nil {
+					for _, record := range failedToSend {
+						t.add(record)
+					}
+				}
+			}
+		case <-t.done:
+			return
+		}
+	}
+}
+
+// cutoff the current record and swap it out with a new record.
+// Sets the timestamp and returns the old record.
+func (t *telemetryRecorder) cutoff() *xray.TelemetryRecord {
+	t.mu.Lock()
+	oldRecord := t.record
+	t.record = newTelemetryRecord()
+	t.mu.Unlock()
+	oldRecord.SetTimestamp(time.Now())
+	return oldRecord
+}
+
+// add to the queue. If queue is full, drop the head of the queue and try again.
+func (t *telemetryRecorder) add(record *xray.TelemetryRecord) {
+	select {
+	case t.queue <- record:
+	default:
+		<-t.queue
+		t.add(record)
+	}
+}
+
+// flush the queue into a slice of records.
+func (t *telemetryRecorder) flush() []*xray.TelemetryRecord {
+	var records []*xray.TelemetryRecord
+	for len(t.queue) > 0 {
+		records = append(records, <-t.queue)
+	}
+	return records
+}
+
+// send the records in batches of batchSize. Returns the error and records it was unable to send
+// if the PutTelemetryRecords call fails.
+func (t *telemetryRecorder) send(records []*xray.TelemetryRecord) ([]*xray.TelemetryRecord, error) {
+	if len(records) > 0 {
+		for i := 0; i < len(records); i += batchSize {
+			endIndex := i + batchSize
+			if endIndex > len(records) {
+				endIndex = len(records)
+			}
+			input := &xray.PutTelemetryRecordsInput{
+				EC2InstanceId:    &t.instanceID,
+				Hostname:         &t.hostname,
+				ResourceARN:      &t.resourceARN,
+				TelemetryRecords: records[i:endIndex],
+			}
+			_, err := t.client.PutTelemetryRecords(input)
+			if err != nil {
+				t.RecordConnectionError(err)
+				return records[i:], err
+			}
+		}
+	}
+	return nil, nil
+}
+
+func (t *telemetryRecorder) RecordSegmentsReceived(count int) {
+	t.mu.Lock()
+	atomic.AddInt64(t.record.SegmentsReceivedCount, int64(count))
+	t.mu.Unlock()
+	t.recordUpdated.Store(true)
+}
+
+func (t *telemetryRecorder) RecordSegmentsSent(count int) {
+	t.mu.Lock()
+	atomic.AddInt64(t.record.SegmentsSentCount, int64(count))
+	t.mu.Unlock()
+	t.recordUpdated.Store(true)
+}
+
+func (t *telemetryRecorder) RecordSegmentsSpillover(count int) {
+	t.mu.Lock()
+	atomic.AddInt64(t.record.SegmentsSpilloverCount, int64(count))
+	t.mu.Unlock()
+	t.recordUpdated.Store(true)
+}
+
+func (t *telemetryRecorder) RecordSegmentsRejected(count int) {
+	t.mu.Lock()
+	atomic.AddInt64(t.record.SegmentsRejectedCount, int64(count))
+	t.mu.Unlock()
+	t.recordUpdated.Store(true)
+}
+
+func (t *telemetryRecorder) RecordConnectionError(err error) {
+	var requestFailure awserr.RequestFailure
+	if ok := errors.As(err, &requestFailure); ok {
+		statusCode := requestFailure.StatusCode()
+		switch {
```

Comment:
If you wanted to make this easier to mentally parse, what you can do is this:

```go
switch statusCode / 100 {
case 5: // xx
case 4: // xx
case 3: // xx
case 2: // xx
case 1: // xx
}
```
I find it easier to mentally understand and review but bounded sets are also fine.